### ISSUE TITLE
optimized_clang tarballs are large and fragile

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylladb#20719]
The optimized_clang tarballs are ~1GB in size and are fragile to install.

I asked Claude, it recommended this:

To build an LLVM/Clang tarball that can be shipped and installed to /usr/local, you'll need to follow these general steps:

Download LLVM source code
Configure the build
Compile LLVM and Clang
Create a tarball
Install from the tarball

Here's a more detailed walkthrough:

Download LLVM source code:
First, clone the LLVM project repository:
Copygit clone https://github.com/llvm/llvm-project.git
cd llvm-project

Configure the build:
Create a build directory and configure the project:

mkdir build
cd build
cmake -G "Unix Makefiles" -DLLVM_ENABLE_PROJECTS=clang -DCMAKE_INSTALL_PREFIX=/usr/local -DCMAKE_BUILD_TYPE=Release ../llvm
This configures the build to include Clang and sets the installation prefix to /usr/local.
Compile LLVM and Clang:

make -j$(nproc)
This will compile LLVM and Clang using all available CPU cores.
Create a tarball:
Instead of installing directly, we'll create a tarball:

make install DESTDIR=/tmp/llvm-install
cd /tmp/llvm-install
tar czf llvm-clang-install.tar.gz usr
This creates a tarball containing the entire installation structure.
Install from the tarball:
To install on the target system:

sudo tar xzf llvm-clang-install.tar.gz -C /
This will extract the contents to /usr/local on the target system.

Remember that building LLVM and Clang requires significant disk space and can take a considerable amount of time, depending on your system's capabilities.
Would you like me to elaborate on any part of this process? CopyRetryClaude does not have the ability to run the code it generates yet.AIs DESTDIR honored with ninja?EditYes, DESTDIR is honored when using Ninja as the build system for LLVM/Clang. The DESTDIR mechanism is typically handled by the build system generator (in this case, CMake) rather than the underlying build tool (like Make or Ninja).
To use Ninja instead of Make in the process I outlined earlier, you would make these adjustments:

When configuring the build with CMake, specify Ninja as the generator:
cmake -G Ninja -DLLVM_ENABLE_PROJECTS=clang -DCMAKE_INSTALL_PREFIX=/usr/local -DCMAKE_BUILD_TYPE=Release ../llvm

For the build step, use Ninja instead of Make:
ninja

When creating the installation for the tarball, use Ninja with DESTDIR:
DESTDIR=/tmp/llvm-install ninja install

The rest of the process (creating the tarball and installing from it) remains the same.
Ninja will respect the DESTDIR variable just like Make does. This allows you to stage the installation in a temporary directory before creating the tarball, ensuring that the final package has the correct directory structure for installation in /usr/local on the target system.

**Comment by mykaul** [comment]
Claude is suggesting a .tar.gz and not zstd? What's wrong with its training?

**Comment by syuu1228** [comment]
I think this is closed by f6fe4d9e73cf993a343f6c6868a903d43c41f66f.
Now it can `sudo tar xzf llvm-clang-install.tar.gz -C /` to install, and reduced size of the optimized_clang tarball:
```
$ ls -lh clang_build/optimized_clang_18.1.6_fedora-40-20240924.x86_64.tar.gz 
-rw-r--r-- 1 syuu docker 72M Sep 24 15:50 clang_build/optimized_clang_18.1.6_fedora-40-20240924.x86_64.tar.gz
```
Now it is less than 100MB.

**Comment by avikivity** [comment]
Closed by https://github.com/scylladb/scylladb/commit/f6fe4d9e73cf993a343f6c6868a903d43c41f66f
